pybop/optimisers/_random_search.py

```diff
@@ -0,0 +1,129 @@
+import numpy as np
+from pints import PopulationBasedOptimiser
+
+
+class RandomSearchImpl(PopulationBasedOptimiser):
+    """
+    Random Search (RS) optimisation algorithm.
+    This algorithm explores the parameter space by randomly sampling points.
+
+    The algorithm is simple:
+    1. Initialise a population of solutions.
+    2. At each iteration, generate new random solutions within boundaries.
+    3. Evaluate the quality/fitness of the solutions.
+    4. Update the best solution found so far.
+
+    Parameters:
+    - population_size (optional): Number of solutions to evaluate per iteration.
+
+    References:
+    - The Random Search algorithm implemented in this work is based on principles outlined
+    in "Introduction to Stochastic Search and Optimization: Estimation, Simulation, and
+    Control" by Spall, J. C. (2003).
+    The implementation leverages the pints library framework, which provides tools for
+    population-based optimization methods.
+    """
+
+    def __init__(self, x0, sigma0=0.05, boundaries=None):
+        # Problem dimensionality
+        self._dim = len(x0)  # Initialize _dim first
+
+        super().__init__(x0, sigma0, boundaries=boundaries)
+
+        # Optimisation parameters
+        self._n = self._population_size
+        self._iterations = 0
+
+        # Initialise population
+        self._candidates = (
+            np.random.uniform(
+                low=self._boundaries.lower(),
+                high=self._boundaries.upper(),
+                size=(self._n, self._dim),
+            )
+            if self._boundaries
+            else np.random.normal(self._x0, self._sigma0, size=(self._n, self._dim))
+        )
+
+        # Initialise best solution
+        self._x_best = np.copy(x0)
+        self._f_best = np.inf
+        self._running = False
+        self._ready_for_tell = False
+
+    def ask(self):
+        """
+        Returns a list of next points in the parameter-space
+        to evaluate from the optimiser.
+        """
+        self._ready_for_tell = True
+        self._running = True
+
+        # Generate random solutions within the boundaries
+        if self._boundaries:
+            self._candidates = np.random.uniform(
+                low=self._boundaries.lower(),
+                high=self._boundaries.upper(),
+                size=(self._n, self._dim),
+            )
+        else:
+            self._candidates = np.random.normal(
+                self._x0, self._sigma0, size=(self._n, self._dim)
+            )
+        return self.clip_candidates(self._candidates)
```

Comment:
Here's an updated version that doesn't duplicate the clipping in the situation that we have `self._boundaries`:

```suggestion
        if self._boundaries:
            self._candidates = np.random.uniform(
                low=self._boundaries.lower(),
                high=self._boundaries.upper(),
                size=(self._n, self._dim),
            )
            return self._candidates

        self._candidates = np.random.normal(
            self._x0, self._sigma0, size=(self._n, self._dim)
        )
        return self.clip_candidates(self._candidates)
```